python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -21,177 +21,842 @@
+            names.append(str(requested))
+
+    return names
+
+
+def _apply_astropy_metadata(astropy_table: atable.Table, metadata: Dict) -> None:
+    """Apply any astropy metadata from the schema metadata.
+
+    Parameters
+    ----------
+    astropy_table : `astropy.table.Table`
+        Table to apply metdata
+    metadata : `dict` [`bytes`]
+        Metadata dict.
+    """
+    from astropy.table import meta
+
+    meta_yaml = metadata.get(b"table_meta_yaml", None)
+    if meta_yaml:
+        meta_yaml = meta_yaml.decode("UTF8").split("\n")
+        meta_hdr = meta.get_header_from_yaml(meta_yaml)
+
+        # Set description, format, unit, meta from the column
+        # metadata that was serialized with the table.
+        header_cols = {x["name"]: x for x in meta_hdr["datatype"]}
+        for col in astropy_table.columns.values():
+            for attr in ("description", "format", "unit", "meta"):
+                if attr in header_cols[col.name]:
+                    setattr(col, attr, header_cols[col.name][attr])
+
+
+def _arrow_string_to_numpy_dtype(
+    schema: pa.Schema, name: str, numpy_column: np.ndarray | None = None, default_length: int = 10
+) -> str:
+    """Get the numpy dtype string associated with an arrow column.
+
+    Parameters
+    ----------
+    schema : `pyarrow.Schema`
+        Arrow table schema.
+    name : `str`
+        Column name.
+    numpy_column : `numpy.ndarray`, optional
+        Column to determine numpy string dtype.
+    default_length : `int`, optional
+        Default string length when not in metadata or can be inferred
+        from column.
+
+    Returns
+    -------
+    dtype_str : `str`
+        Numpy dtype string.
+    """
+    # Special-case for string and binary columns
+    md_name = f"lsst::arrow::len::{name}"
+    strlen = default_length
+    metadata = schema.metadata if schema.metadata is not None else {}
+    if md_name.encode("UTF-8") in metadata:
+        # String/bytes length from header.
+        strlen = int(schema.metadata[md_name.encode("UTF-8")])
+    elif numpy_column is not None:
+        if len(numpy_column) > 0:
+            strlen = max(len(row) for row in numpy_column)
+
+    dtype = f"U{strlen}" if schema.field(name).type == pa.string() else f"|S{strlen}"
+
+    return dtype
+
+
+def _append_numpy_string_metadata(metadata: Dict[bytes, str], name: str, dtype: np.dtype) -> None:
+    """Append numpy string length keys to arrow metadata.
+
+    Parameters
+    ----------
+    metadata : `dict` [`bytes`, `str`]
+        Metadata dictionary; modified in place.
+    name : `str`
+        Column name.
+    dtype : `np.dtype`
+        Numpy dtype.
+    """
+    import numpy as np
+
+    if dtype.type is np.str_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize // 4)
+    elif dtype.type is np.bytes_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize)
```

Comment:
Is this called with a non-string dtype ever?

Reply:
Yes, and then this is a no-op.  Is that a problem?

Reply:
I wanted to know if the dangling else was okay. Maybe add a note to the docstring saying that all column types are handled but only string types are modified?